File: packages/cli/src/services/untrack.ts
import { readdir, rm } from "node:fs/promises";
import { dirname, join, posix } from "node:path";
import type { ResolvedSyncConfigEntry } from "#app/config/sync-schema.ts";
import { DotweaveError } from "#app/lib/error.ts";
import {
  getPathStats,
  listDirectoryEntries,
  removePathAtomically,
} from "#app/lib/filesystem.ts";
import { isPathEqualOrNested } from "#app/lib/path.ts";
import {
  buildSyncConfigDocument,
  writeValidatedSyncConfig,
} from "./config-file.ts";
import {
  collectArtifactProfiles,
  isSecretArtifactPath,
  resolveArtifactRelativePath,
} from "./repo-artifacts.ts";
import { loadWritableSyncConfig } from "./sync-context.ts";
import { resolveTrackedEntry } from "./sync-paths.ts";

export type UntrackRequest = Readonly<{
  target: string;
}>;

export type UntrackResult = Readonly<{
  localPath: string;
  plainArtifactCount: number;
  repoPath: string;
  secretArtifactCount: number;
}>;

const collectRepoArtifactCounts = async (
  targetPath: string,
  counts: {
    plain: number;
    secret: number;
  },
  relativePath: string,
) => {
  const stats = await getPathStats(targetPath);

  if (stats === undefined) {
    return;
  }

  if (stats.isDirectory()) {
    counts.plain += 1;

    const entries = await listDirectoryEntries(targetPath);

    for (const entry of entries) {
      await collectRepoArtifactCounts(
        join(targetPath, entry.name),
        counts,
        posix.join(relativePath, entry.name),
      );
    }

    return;
  }

  if (isSecretArtifactPath(relativePath)) {
    counts.secret += 1;
  } else {
    counts.plain += 1;
  }
};

const collectEntryArtifactCounts = async (
  syncDirectory: string,
  entry: ResolvedSyncConfigEntry,
) => {
  const artifactsRoot = syncDirectory;
  const counts = {
    plain: 0,
    secret: 0,
  };
  const artifactProfiles = collectArtifactProfiles([entry]);

  for (const profile of artifactProfiles) {
    const plainRelativePath = resolveArtifactRelativePath({
      category: "plain",
      profile,
      repoPath: entry.repoPath,
    });

    await collectRepoArtifactCounts(
      join(artifactsRoot, ...plainRelativePath.split("/")),
      counts,
      plainRelativePath,
    );

    if (entry.kind !== "directory") {
      const secretRelativePath = resolveArtifactRelativePath({
        category: "secret",
        profile,
        repoPath: entry.repoPath,
      });

      await collectRepoArtifactCounts(
        join(artifactsRoot, ...secretRelativePath.split("/")),
        counts,
        secretRelativePath,
      );

      const symlinkRelativePath = resolveArtifactRelativePath({
        category: "plain",
        kind: "symlink",
        profile,
        repoPath: entry.repoPath,
      });

      await collectRepoArtifactCounts(
        join(artifactsRoot, ...symlinkRelativePath.split("/")),
        counts,
        symlinkRelativePath,
      );
    }
  }

  return {
    plainArtifactCount: counts.plain,
    secretArtifactCount: counts.secret,
  };
};

const pruneEmptyParentDirectories = async (
  startPath: string,
  rootPath: string,
) => {
  let currentPath = startPath;

  while (
    isPathEqualOrNested(currentPath, rootPath) &&
    currentPath !== rootPath
  ) {
    const stats = await getPathStats(currentPath);

    if (stats === undefined) {
      currentPath = dirname(currentPath);
      continue;
    }

    if (!stats.isDirectory()) {
      break;
    }

    const entries = await readdir(currentPath);

    if (entries.length > 0) {
      break;
    }

    await rm(currentPath, { force: true, recursive: true });
    currentPath = dirname(currentPath);
  }
};

const removeTrackedEntryArtifacts = async (
  syncDirectory: string,
  entry: ResolvedSyncConfigEntry,
) => {
  const artifactsRoot = syncDirectory;
  const artifactProfiles = collectArtifactProfiles([entry]);

  for (const profile of artifactProfiles) {
    const plainPath = join(
      artifactsRoot,
      ...resolveArtifactRelativePath({
        category: "plain",
        profile,
        repoPath: entry.repoPath,
      }).split("/"),
    );

    await removePathAtomically(plainPath);
    await pruneEmptyParentDirectories(dirname(plainPath), artifactsRoot);

    if (entry.kind !== "directory") {
      const secretPath = join(
        artifactsRoot,
        ...resolveArtifactRelativePath({
          category: "secret",
          profile,
          repoPath: entry.repoPath,
        }).split("/"),
      );

      await removePathAtomically(secretPath);
      await pruneEmptyParentDirectories(dirname(secretPath), artifactsRoot);

      const symlinkPath = join(
        artifactsRoot,
        ...resolveArtifactRelativePath({
          category: "plain",
          kind: "symlink",
          profile,
          repoPath: entry.repoPath,
        }).split("/"),
      );

      await removePathAtomically(symlinkPath);
      await pruneEmptyParentDirectories(dirname(symlinkPath), artifactsRoot);
    }
  }
};

export const untrackTarget = async (
  request: UntrackRequest,
  cwd: string,
): Promise<UntrackResult> => {
  const target = request.target.trim();

  if (target.length === 0) {
    throw new DotweaveError("Target path is required.");
  }

  const { config, context, syncDirectory } = await loadWritableSyncConfig();
  const entry = resolveTrackedEntry(
    target,
    config.entries,
    cwd,
    context.homeDirectory,
  );

  if (entry === undefined) {
    throw new DotweaveError(`No tracked sync entry matches: ${target}`);
  }

  const { plainArtifactCount, secretArtifactCount } =
    await collectEntryArtifactCounts(syncDirectory, entry);
  const nextConfig = buildSyncConfigDocument({
    ...config,
    entries: config.entries.filter((configEntry) => {
      return configEntry.repoPath !== entry.repoPath;
    }),
  });

  await writeValidatedSyncConfig(syncDirectory, nextConfig);
  await removeTrackedEntryArtifacts(syncDirectory, entry);

  return {
    localPath: entry.localPath,
    plainArtifactCount,
    repoPath: entry.repoPath,
    secretArtifactCount,
  };
};

File: packages/cli/lib/src/services/untrack.dart
import 'dart:io';

import 'package:dotweave/src/config/sync_schema.dart';
import 'package:dotweave/src/services/config_file.dart';
import 'package:dotweave/src/services/repo_artifact_path.dart';
import 'package:dotweave/src/services/repo_artifacts.dart';
import 'package:dotweave/src/services/sync_context.dart';
import 'package:dotweave/src/services/sync_paths.dart';
import 'package:dotweave/src/util/error.dart';
import 'package:dotweave/src/util/filesystem.dart';
import 'package:dotweave/src/util/path_util.dart';
import 'package:path/path.dart' as p;

// Mirror of `services/untrack.ts`: removal of a tracked entry from the
// manifest plus cleanup of its repository artifacts.

/// Mirror of the TS `UntrackRequest` readonly object.
class UntrackRequest {
  const UntrackRequest({required this.target});

  final String target;
}

/// Mirror of the TS `UntrackResult` readonly object.
class UntrackResult {
  const UntrackResult({
    required this.localPath,
    required this.plainArtifactCount,
    required this.repoPath,
    required this.secretArtifactCount,
  });

  final String localPath;
  final int plainArtifactCount;
  final String repoPath;
  final int secretArtifactCount;

  @override
  bool operator ==(Object other) {
    return other is UntrackResult &&
        other.localPath == localPath &&
        other.plainArtifactCount == plainArtifactCount &&
        other.repoPath == repoPath &&
        other.secretArtifactCount == secretArtifactCount;
  }

  @override
  int get hashCode =>
      Object.hash(localPath, plainArtifactCount, repoPath, secretArtifactCount);

  @override
  String toString() {
    return 'UntrackResult(localPath: $localPath, '
        'plainArtifactCount: $plainArtifactCount, repoPath: $repoPath, '
        'secretArtifactCount: $secretArtifactCount)';
  }
}

/// Optional overrides for [untrackTarget], standing in for the vitest module
/// mocks used by `untrack.test.ts`.
class UntrackDependencies {
  const UntrackDependencies({
    this.buildSyncConfigDocument,
    this.loadWritableSyncConfig,
    this.resolveTrackedEntry,
    this.writeValidatedSyncConfig,
  });

  final RawSyncConfig Function(ResolvedSyncConfig config)?
  buildSyncConfigDocument;
  final Future<WritableSyncConfig> Function()? loadWritableSyncConfig;
  final ResolvedSyncConfigEntry? Function(
    String target,
    List<ResolvedSyncConfigEntry> entries,
    String cwd,
    String homeDirectory,
  )?
  resolveTrackedEntry;
  final Future<void> Function(String syncDirectory, RawSyncConfig config)?
  writeValidatedSyncConfig;
}

/// Mirror of the mutable `counts` object threaded through the TS
/// `collectRepoArtifactCounts`.
class _ArtifactCounts {
  int plain = 0;
  int secret = 0;
}

Future<void> _collectRepoArtifactCounts(
  String targetPath,
  _ArtifactCounts counts,
  String relativePath,
) async {
  final stats = await getPathStats(targetPath);

  if (stats == null) {
    return;
  }

  if (stats.isDirectory) {
    counts.plain += 1;

    final entries = await listDirectoryEntries(targetPath);

    for (final entry in entries) {
      await _collectRepoArtifactCounts(
        p.join(targetPath, entry.name),
        counts,
        p.posix.join(relativePath, entry.name),
      );
    }

    return;
  }

  if (isSecretArtifactPath(relativePath)) {
    counts.secret += 1;
  } else {
    counts.plain += 1;
  }
}

Future<({int plainArtifactCount, int secretArtifactCount})>
_collectEntryArtifactCounts(
  String syncDirectory,
  ResolvedSyncConfigEntry entry,
) async {
  final artifactsRoot = syncDirectory;
  final counts = _ArtifactCounts();
  final artifactProfiles = collectArtifactProfiles(
    entries: <ResolvedSyncConfigEntry>[entry],
  );

  for (final profile in artifactProfiles) {
    final plainRelativePath = resolveArtifactRelativePath(
      category: 'plain',
      profile: profile,
      repoPath: entry.repoPath,
    );

    await _collectRepoArtifactCounts(
      p.joinAll([artifactsRoot, ...plainRelativePath.split('/')]),
      counts,
      plainRelativePath,
    );

    if (entry.kind != 'directory') {
      final secretRelativePath = resolveArtifactRelativePath(
        category: 'secret',
        profile: profile,
        repoPath: entry.repoPath,
      );

      await _collectRepoArtifactCounts(
        p.joinAll([artifactsRoot, ...secretRelativePath.split('/')]),
        counts,
        secretRelativePath,
      );

      final symlinkRelativePath = resolveArtifactRelativePath(
        category: 'plain',
        kind: 'symlink',
        profile: profile,
        repoPath: entry.repoPath,
      );

      await _collectRepoArtifactCounts(
        p.joinAll([artifactsRoot, ...symlinkRelativePath.split('/')]),
        counts,
        symlinkRelativePath,
      );
    }
  }

  return (plainArtifactCount: counts.plain, secretArtifactCount: counts.secret);
}

Future<void> _pruneEmptyParentDirectories(
  String startPath,
  String rootPath,
) async {
  var currentPath = startPath;

  while (isPathEqualOrNested(currentPath, rootPath) &&
      currentPath != rootPath) {
    final stats = await getPathStats(currentPath);

    if (stats == null) {
      currentPath = p.dirname(currentPath);
      continue;
    }

    if (!stats.isDirectory) {
      break;
    }

    final entries = await listDirectoryEntries(currentPath);

    if (entries.isNotEmpty) {
      break;
    }

    try {
      await Directory(currentPath).delete(recursive: true);
    } on PathNotFoundException {
      // Mirrors node:fs `rm` with `force: true`: missing paths are ignored.
    }
    currentPath = p.dirname(currentPath);
  }
}

Future<void> _removeTrackedEntryArtifacts(
  String syncDirectory,
  ResolvedSyncConfigEntry entry,
) async {
  final artifactsRoot = syncDirectory;
  final artifactProfiles = collectArtifactProfiles(
    entries: <ResolvedSyncConfigEntry>[entry],
  );

  for (final profile in artifactProfiles) {
    final plainPath = p.joinAll([
      artifactsRoot,
      ...resolveArtifactRelativePath(
        category: 'plain',
        profile: profile,
        repoPath: entry.repoPath,
      ).split('/'),
    ]);

    await removePathAtomically(plainPath);
    await _pruneEmptyParentDirectories(p.dirname(plainPath), artifactsRoot);

    if (entry.kind != 'directory') {
      final secretPath = p.joinAll([
        artifactsRoot,
        ...resolveArtifactRelativePath(
          category: 'secret',
          profile: profile,
          repoPath: entry.repoPath,
        ).split('/'),
      ]);

      await removePathAtomically(secretPath);
      await _pruneEmptyParentDirectories(p.dirname(secretPath), artifactsRoot);

      final symlinkPath = p.joinAll([
        artifactsRoot,
        ...resolveArtifactRelativePath(
          category: 'plain',
          kind: 'symlink',
          profile: profile,
          repoPath: entry.repoPath,
        ).split('/'),
      ]);

      await removePathAtomically(symlinkPath);
      await _pruneEmptyParentDirectories(p.dirname(symlinkPath), artifactsRoot);
    }
  }
}

Future<UntrackResult> untrackTarget(
  UntrackRequest request,
  String cwd, [
  UntrackDependencies dependencies = const UntrackDependencies(),
]) async {
  final effectiveLoadWritableSyncConfig =
      dependencies.loadWritableSyncConfig ?? loadWritableSyncConfig;
  final effectiveResolveTrackedEntry =
      dependencies.resolveTrackedEntry ?? resolveTrackedEntry;
  final effectiveBuildSyncConfigDocument =
      dependencies.buildSyncConfigDocument ?? buildSyncConfigDocument;
  final effectiveWriteValidatedSyncConfig =
      dependencies.writeValidatedSyncConfig ?? writeValidatedSyncConfig;

  final target = request.target.trim();

  if (target.isEmpty) {
    throw DotweaveError('Target path is required.');
  }

  final writable = await effectiveLoadWritableSyncConfig();
  final config = writable.config;
  final context = writable.context;
  final syncDirectory = writable.syncDirectory;
  final entry = effectiveResolveTrackedEntry(
    target,
    config.entries,
    cwd,
    context.homeDirectory,
  );

  if (entry == null) {
    throw DotweaveError('No tracked sync entry matches: $target');
  }

  final counts = await _collectEntryArtifactCounts(syncDirectory, entry);
  final nextConfig = effectiveBuildSyncConfigDocument(
    ResolvedSyncConfig(
      age: config.age,
      entries: [
        for (final configEntry in config.entries)
          if (configEntry.repoPath != entry.repoPath) configEntry,
      ],
      profiles: config.profiles,
      repositoryFormat: config.repositoryFormat,
      version: config.version,
    ),
  );

  await effectiveWriteValidatedSyncConfig(syncDirectory, nextConfig);
  await _removeTrackedEntryArtifacts(syncDirectory, entry);

  return UntrackResult(
    localPath: entry.localPath,
    plainArtifactCount: counts.plainArtifactCount,
    repoPath: entry.repoPath,
    secretArtifactCount: counts.secretArtifactCount,
  );
}
